File: backend/src/clawbuddy/routers/skills.py
"""Skills router — upload, list, delete, and rebuild skill images.

Replaces: apps/api/src/routes/skills.ts
"""

from __future__ import annotations

import json
from typing import Any

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from clawbuddy.db.session import get_db
from clawbuddy.lib.responses import fail, ok
from clawbuddy.lib.sse import create_sse_stream
from clawbuddy.services.skill import skill_service

router = APIRouter(tags=["skills"])


@router.post("/skills/upload")
async def upload_skill(
    body: dict[str, Any],
    db: AsyncSession = Depends(get_db),
):
    """Upload a .skill file.

    If it has an installation script, the endpoint streams build logs via SSE.
    """
    has_installation = bool(body.get("installation"))

    if has_installation:
        # Stream build logs via SSE
        async def handler(emit):
            def on_build_log(line: str) -> None:
                import asyncio

                # Use fire-and-forget since on_build_log is sync
                try:
                    loop = asyncio.get_running_loop()
                    loop.create_task(
                        emit("build_log", {"data": line})
                    )
                except RuntimeError:
                    pass

            result = await skill_service.upload_skill(
                json.dumps(body), db, on_build_log
            )

            if result.success:
                await emit("complete", {
                    "success": True,
                    "slug": result.slug,
                })
            else:
                await emit("error", {
                    "success": False,
                    "error": result.error,
                    "logs": result.logs,
                })

        return create_sse_stream(handler)

    # No installation script — simple JSON response
    result = await skill_service.upload_skill(json.dumps(body), db)
    if not result.success:
        return fail(
            result.error or "Upload failed",
            status_code=400,
        )
    return ok({"slug": result.slug})


@router.get("/skills")
async def list_skills(db: AsyncSession = Depends(get_db)):
    """List all installed skills."""
    skills = await skill_service.list_skills(db)
    return ok([
        {
            "id": s.id,
            "slug": s.slug,
            "name": s.name,
            "description": s.description,
            "icon": s.icon,
            "category": s.category,
            "version": s.version,
            "skillType": s.skill_type,
            "source": s.source,
        }
        for s in skills
    ])


@router.delete("/skills/{slug}")
async def delete_skill(
    slug: str,
    db: AsyncSession = Depends(get_db),
):
    """Delete a skill."""
    result = await skill_service.delete_skill(slug, db)
    if not result.success:
        return fail(result.error or "Delete failed", status_code=400)
    return ok({"deleted": True})


@router.post("/skills/rebuild-image")
async def rebuild_skill_image(
    body: dict[str, Any],
    db: AsyncSession = Depends(get_db),
):
    """Force rebuild the skill Docker image for a workspace."""
    workspace_id = body.get("workspaceId")
    if not workspace_id:
        return fail("workspaceId is required", status_code=400)

    async def handler(emit):
        try:
            from clawbuddy.services.image_builder import image_builder_service
            from clawbuddy.services.sandbox import sandbox_service

            await image_builder_service.invalidate_images()
            await emit("build_log", {"data": "Invalidated old images"})

            async def on_log_async(line: str) -> None:
                await emit("build_log", {"data": line})

            # on_log needs to be sync for the image builder, so use a wrapper
            import asyncio

            def on_log(line: str) -> None:
                try:
                    loop = asyncio.get_running_loop()
                    loop.create_task(emit("build_log", {"data": line}))
                except RuntimeError:
                    pass

            tag = await image_builder_service.get_or_build_image(
                workspace_id, db, on_log
            )

            await emit("build_log", {"data": "Stopping workspace container..."})
            try:
                await sandbox_service.stop_workspace_container(
                    workspace_id, db
                )
            except Exception:
                pass

            await emit("complete", {"success": True, "image": tag})
        except Exception as exc:
            message = str(exc) or "Failed to rebuild image"
            await emit("error", {"success": False, "error": message})

    return create_sse_stream(handler)

File: apps/api/src/routes/skills.ts
import { Hono } from 'hono'
import { streamSSE } from 'hono/streaming'
import { skillService } from '../services/skill.service.js'
import { imageBuilderService } from '../services/image-builder.service.js'
import { sandboxService } from '../services/sandbox.service.js'
import { logger } from '../lib/logger.js'
import { parseSkillSource } from '../capabilities/skill-parser.js'

const app = new Hono()

/**
 * Upload a skill source file. If it has an installation script, the endpoint
 * streams build logs via SSE so the frontend can show progress.
 */
app.post('/skills/upload', async (c) => {
  const body = await c.req.json()
  const content = typeof body?.content === 'string' ? body.content : null
  const fileName = typeof body?.filename === 'string' ? body.filename : undefined

  if (!content) {
    return c.json({ success: false, error: 'Skill content is required' }, 400)
  }

  // Check if the skill has an installation script to determine response type
  let hasInstallation = false
  try {
    hasInstallation = !!parseSkillSource(content).skill.installation
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Invalid skill data'
    return c.json({ success: false, error: message }, 400)
  }

  if (hasInstallation) {
    // Stream build logs via SSE
    return streamSSE(c, async (stream) => {
      const result = await skillService.uploadSkill(content, {
        fileName,
        onBuildLog: (line) => {
          stream.writeSSE({ event: 'build_log', data: line })
        },
      })

      if (result.success) {
        await stream.writeSSE({
          event: 'complete',
          data: JSON.stringify({ success: true, slug: result.slug }),
        })
      } else {
        await stream.writeSSE({
          event: 'error',
          data: JSON.stringify({
            success: false,
            error: result.error,
            logs: result.logs,
          }),
        })
      }
    })
  }

  // No installation script — simple JSON response
  const result = await skillService.uploadSkill(content, { fileName })
  if (!result.success) {
    return c.json({ success: false, error: result.error, logs: result.logs }, 400)
  }
  return c.json({ success: true, data: { slug: result.slug } }, 201)
})

/**
 * List all installed skills.
 */
app.get('/skills', async (c) => {
  const skills = await skillService.listSkills()
  return c.json({ success: true, data: skills })
})

/**
 * Delete a skill.
 */
app.delete('/skills/:slug', async (c) => {
  const { slug } = c.req.param()
  const result = await skillService.deleteSkill(slug)
  if (!result.success) {
    return c.json({ success: false, error: result.error }, 400)
  }
  return c.json({ success: true, data: { deleted: true } })
})

/**
 * Force rebuild the skill Docker image for a workspace.
 */
app.post('/skills/rebuild-image', async (c) => {
  const { workspaceId, clearCache } = await c.req.json()
  if (!workspaceId) {
    return c.json({ success: false, error: 'workspaceId is required' }, 400)
  }

  return streamSSE(c, async (stream) => {
    try {
      if (clearCache) {
        await imageBuilderService.invalidateImages()
        await stream.writeSSE({ event: 'build_log', data: 'Cleared cached images' })
      } else {
        await stream.writeSSE({
          event: 'build_log',
          data: 'Reusing cached layers when available...',
        })
      }

      const tag = await imageBuilderService.getOrBuildImage(workspaceId, (line) => {
        stream.writeSSE({ event: 'build_log', data: line })
      })

      // Stop the running container so the next execution uses the new image
      await stream.writeSSE({ event: 'build_log', data: 'Stopping workspace container...' })
      await sandboxService.stopWorkspaceContainer(workspaceId).catch((err) =>
        logger.warn('[Skills] Failed to stop workspace container after build', {
          workspaceId,
          error: err instanceof Error ? err.message : String(err),
        }),
      )

      await stream.writeSSE({
        event: 'complete',
        data: JSON.stringify({ success: true, image: tag }),
      })
    } catch (err) {
      const message = err instanceof Error ? err.message : 'Failed to rebuild image'
      await stream.writeSSE({
        event: 'error',
        data: JSON.stringify({ success: false, error: message }),
      })
    }
  })
})

export default app
